File: ui/src/api/hooks/transactions.ts
import { useInfiniteQuery, useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { api, type Transaction, type TxnFilterInput, type NewTransaction, type CsvImportMapping, type ImportResult, type TxnPatch, type UpdateTxnResult, type CategoryDto, type CategoryWithSpending, type CategoryGroup, type RuleWithCategory, type SplitInputDto } from "../openapiClient";
import { unwrap } from "../openapiClient";
import { isBackendAvailable } from "../../utils/runtime";
import { invalidateDomains } from "../invalidation";

const DEFAULT_FILTER: TxnFilterInput = {
  accountId: null,
  limit: null,
  offset: null,
  search: null,
  filterPreset: null,
  startDate: null,
  endDate: null,
};

export function useTransactions(filter: TxnFilterInput = DEFAULT_FILTER) {
  return useQuery<Transaction[]>({
    queryKey: ["transactions", filter],
    queryFn: async () => {
      return unwrap(api.listTransactions(filter));
    },
    enabled: isBackendAvailable(),
  });
}

/** Page size for the paginated transactions list. */
export const TXN_PAGE_SIZE = 50;

/**
 * Paginated transactions via infinite query. Filters/sort/search all flow
 * through `filter`; changing any of them starts a fresh paged query (the filter
 * is part of the query key), so older transactions stay reachable via
 * `fetchNextPage` without ever loading thousands of rows at once.
 */
export function useInfiniteTransactions(
  filter: Omit<TxnFilterInput, "limit" | "offset">,
) {
  return useInfiniteQuery({
    queryKey: ["transactions-infinite", filter],
    initialPageParam: 0,
    queryFn: async ({ pageParam }) => {
      return unwrap(api.listTransactions({
        ...filter,
        limit: TXN_PAGE_SIZE,
        offset: pageParam * TXN_PAGE_SIZE,
      } as TxnFilterInput));
    },
    getNextPageParam: (lastPage, allPages) =>
      lastPage.length < TXN_PAGE_SIZE ? undefined : allPages.length,
    enabled: isBackendAvailable(),
  });
}

export function useCreateTransaction() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (input: NewTransaction) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      return unwrap(api.createTransaction(input));
    },
    onSuccess: () => {
      invalidateDomains(qc, "transactions");
    },
  });
}

export function useImportCsv() {
  const qc = useQueryClient();
  return useMutation<ImportResult, Error, { path: string; account_id: string; mapping: CsvImportMapping }>({
    mutationFn: async (args) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      return unwrap(api.importCsv(args.path, args.account_id, args.mapping));
    },
    onSuccess: () => {
      // A CSV commit touches the ledger, account balances, and import state
      // (mapping + any cached speculative preview).
      invalidateDomains(qc, "importCommit");
    },
  });
}

export function useUpdateTransaction() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ id, patch }: { id: string; patch: TxnPatch }) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      return (await unwrap(api.updateTransaction(id, patch))) as UpdateTxnResult;
    },
    onSuccess: () => {
      invalidateDomains(qc, "transactions");
    },
  });
}

export function useDeleteTransaction() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (id: string) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      await unwrap(api.deleteTransaction(id));
    },
    onSuccess: () => {
      invalidateDomains(qc, "transactions");
    },
  });
}

/** Mark a flagged anomaly as reviewed-and-fine (dismiss) or restore it. The
 *  detector won't re-flag a dismissed charge, so the anomaly feed stays clean. */
export function useSetAnomalyDismissed() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ txnId, dismissed }: { txnId: string; dismissed: boolean }) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      await unwrap(api.setAnomalyDismissed(txnId, dismissed));
    },
    onSuccess: () => {
      invalidateDomains(qc, "transactions");
    },
  });
}

export function useCreateRule() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ pattern, categoryId }: { pattern: string; categoryId: string }) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      return unwrap(api.createRule(pattern, categoryId));
    },
    onSuccess: () => {
      invalidateDomains(qc, "rules");
    },
  });
}

export function useSetTransactionOwner() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ transactionId, memberId }: { transactionId: string; memberId: string | null }) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      await unwrap(api.setTransactionOwner(transactionId, memberId));
    },
    // Attribution changes per-member cashflow — refresh transactions and metrics.
    onSuccess: () => {
      invalidateDomains(qc, "transactions");
      void qc.invalidateQueries();
    },
  });
}

export function useCategories() {
  return useQuery<CategoryDto[]>({
    queryKey: ["categories"],
    queryFn: async () => {
      return unwrap(api.listCategories());
    },
    enabled: isBackendAvailable(),
  });
}

export function useCategoriesWithSpending() {
  return useQuery<CategoryWithSpending[]>({
    queryKey: ["categories-with-spending"],
    queryFn: async () => {
      return unwrap(api.listCategoriesWithSpending());
    },
    enabled: isBackendAvailable(),
  });
}

export function useSetCategorySpendingType() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ id, spendingType }: { id: string; spendingType: string | null }) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      await unwrap(api.setCategorySpendingType(id, spendingType));
    },
    onSuccess: () => {
      invalidateDomains(qc, "categories");
    },
  });
}

const invalidateCategoryQueries = (qc: ReturnType<typeof useQueryClient>) =>
  invalidateDomains(qc, "categories");

export function useCreateCategory() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ label, groupId, color }: { label: string; groupId: string | null; color: string }) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      return unwrap(api.createCategory(label, groupId, color));
    },
    onSuccess: () => invalidateCategoryQueries(qc),
  });
}

export function useRenameCategory() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ id, label }: { id: string; label: string }) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      await unwrap(api.renameCategory(id, label));
    },
    onSuccess: () => invalidateCategoryQueries(qc),
  });
}

export function useArchiveCategory() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (id: string) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      await unwrap(api.archiveCategory(id));
    },
    onSuccess: () => {
      // The categories domain already includes rules (archiving a category can
      // disable its rules).
      invalidateCategoryQueries(qc);
    },
  });
}

export function useSetCategoryGuidance() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ id, guidance }: { id: string; guidance: string | null }) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      await unwrap(api.setCategoryGuidance(id, guidance));
    },
    onSuccess: () => invalidateCategoryQueries(qc),
  });
}

export function useUpdateCategoryColor() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ id, color }: { id: string; color: string }) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      await unwrap(api.updateCategoryColor(id, color));
    },
    onSuccess: () => {
      invalidateDomains(qc, "categories");
    },
  });
}

export function useCategoryGroups() {
  return useQuery<CategoryGroup[]>({
    queryKey: ["category-groups"],
    queryFn: async () => {
      return unwrap(api.listCategoryGroups());
    },
    enabled: isBackendAvailable(),
  });
}

export function useCreateCategoryGroup() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ label, hint }: { label: string; hint?: string | null }) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      return unwrap(api.createCategoryGroup(label, hint ?? null));
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["category-groups"] });
    },
  });
}

export function useSetCategoryGroup() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ categoryId, groupId }: { categoryId: string; groupId: string }) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      await unwrap(api.setCategoryGroup(categoryId, groupId));
    },
    onSuccess: () => invalidateCategoryQueries(qc),
  });
}

export function useSetCategoryRollover() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ id, rolloverEnabled }: { id: string; rolloverEnabled: boolean }) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      await unwrap(api.setCategoryRollover(id, rolloverEnabled));
    },
    onSuccess: () => {
      // Rollover flips carryover for the next month, so budget envelopes also change.
      invalidateDomains(qc, "categories");
      invalidateDomains(qc, "budget");
      qc.invalidateQueries({ queryKey: ["budget-envelopes"] });
    },
  });
}

export function useRulesWithCategories() {
  return useQuery<RuleWithCategory[]>({
    queryKey: ["rules"],
    queryFn: async () => {
      return unwrap(api.listRulesWithCategories());
    },
    enabled: isBackendAvailable(),
  });
}

export function useToggleRule() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ id, enabled }: { id: string; enabled: boolean }) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      await unwrap(api.toggleRule(id, enabled));
    },
    onSuccess: () => {
      invalidateDomains(qc, "rules");
    },
  });
}

export function useSetTransactionFlags() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ id, isReimbursable, isSplit }: { id: string; isReimbursable: boolean; isSplit: boolean }) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      return unwrap(api.setTransactionFlags(id, isReimbursable, isSplit));
    },
    onSuccess: () => {
      invalidateDomains(qc, "transactions");
    },
  });
}

/** Record the user's verdict on whether a transaction is a transfer between
 *  their own accounts. Sticky — survives re-imports and categorizer re-runs.
 *  The result reports undecided siblings with the same counterparty so the UI
 *  can offer a one-click bulk verdict. */
export function useSetTransactionTransfer() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ id, isTransfer }: { id: string; isTransfer: boolean }) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      return unwrap(api.setTransactionTransfer(id, isTransfer));
    },
    // A transfer verdict moves money in/out of income & spending — every
    // headline number (savings rate, cashflow, budget, inbox) can change.
    onSuccess: () => {
      invalidateDomains(qc, "transactions");
      void qc.invalidateQueries();
    },
  });
}

/** Apply a transfer verdict to every undecided transaction with the same
 *  counterparty (pattern from `useSetTransactionTransfer`'s result). */
export function useApplyTransferVerdictToSimilar() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ pattern, isTransfer }: { pattern: string; isTransfer: boolean }) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      return unwrap(api.applyTransferVerdictToSimilar(pattern, isTransfer));
    },
    onSuccess: () => {
      invalidateDomains(qc, "transactions");
      void qc.invalidateQueries();
    },
  });
}

export function useTransactionSplits(txnId: string | undefined) {
  return useQuery({
    queryKey: ["splits", txnId],
    queryFn: async () => {
      if (!txnId) return [];
      return unwrap(api.getTransactionSplits(txnId));
    },
    enabled: !!txnId && isBackendAvailable(),
  });
}

export function useSetTransactionSplits() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ txnId, splits }: {
      txnId: string;
      splits: Array<{ categoryId: string | null; amountCents: number }>;
    }) => {
      if (!isBackendAvailable()) throw new Error("This action needs a connected FinSight server.");
      await unwrap(api.setTransactionSplits(
        txnId,
        splits.map((s): SplitInputDto => ({ categoryId: s.categoryId, amountCents: s.amountCents }))
      ));
    },
    onSuccess: (_data, vars) => {
      // A split reassigns amounts across categories, so it touches both the
      // ledger and category spending; plus this txn's own split rows.
      invalidateDomains(qc, "transactions", "categories");
      qc.invalidateQueries({ queryKey: ["splits", vars.txnId] });
    },
  });
}

File: crates/finsight-bindings/src/commands/transactions.rs
use crate::error::AppResult;
use crate::AppState;
use finsight_core::models::{NewTransaction, Rule, Transaction, TxnPatch};

pub use finsight_api::commands::transactions::{
    CategoryDto, CategoryWithSpending, CounterpartyVerdict, ProposedRuleDto, RuleWithCategory,
    SearchTxnQueryInput, SpendingBreakdown, SplitInputDto, TransactionSplitDto,
    TransferVerdictResult, TxnFilterInput, UnresolvedCounterpartyDto, UpdateTxnResult,
};

#[tauri::command]
#[specta::specta]
pub async fn list_transactions(
    state: tauri::State<'_, AppState>,
    filter: TxnFilterInput,
) -> AppResult<Vec<Transaction>> {
    finsight_api::commands::transactions::list_transactions(&state.api, filter).await
}

#[tauri::command]
#[specta::specta]
pub async fn create_transaction(
    state: tauri::State<'_, AppState>,
    input: NewTransaction,
) -> AppResult<Transaction> {
    finsight_api::commands::transactions::create_transaction(&state.api, input).await
}

#[tauri::command]
#[specta::specta]
pub async fn update_transaction(
    state: tauri::State<'_, AppState>,
    id: String,
    patch: TxnPatch,
) -> AppResult<UpdateTxnResult> {
    finsight_api::commands::transactions::update_transaction(&state.api, id, patch).await
}

#[tauri::command]
#[specta::specta]
pub async fn delete_transaction(state: tauri::State<'_, AppState>, id: String) -> AppResult<()> {
    finsight_api::commands::transactions::delete_transaction(&state.api, id).await
}

#[tauri::command]
#[specta::specta]
pub async fn create_rule(
    state: tauri::State<'_, AppState>,
    pattern: String,
    category_id: String,
) -> AppResult<Rule> {
    finsight_api::commands::transactions::create_rule(&state.api, pattern, category_id).await
}

/// Attribute a single transaction to one household member, overriding its
/// account's ownership shares for that row's cashflow — for a personal purchase
/// on a joint account. `member_id` None clears the override (revert to account
/// shares). Only flows are affected; balances are per-account.
#[tauri::command]
#[specta::specta]
pub async fn set_transaction_owner(
    state: tauri::State<'_, AppState>,
    transaction_id: String,
    member_id: Option<String>,
) -> AppResult<()> {
    finsight_api::commands::transactions::set_transaction_owner(
        &state.api,
        transaction_id,
        member_id,
    )
    .await
}

#[tauri::command]
#[specta::specta]
pub async fn list_categories(state: tauri::State<'_, AppState>) -> AppResult<Vec<CategoryDto>> {
    finsight_api::commands::transactions::list_categories(&state.api).await
}

#[tauri::command]
#[specta::specta]
pub async fn set_category_spending_type(
    state: tauri::State<'_, AppState>,
    id: String,
    spending_type: Option<String>,
) -> AppResult<()> {
    finsight_api::commands::transactions::set_category_spending_type(
        &state.api,
        id,
        spending_type,
    )
    .await
}

#[tauri::command]
#[specta::specta]
pub async fn get_spending_breakdown(
    state: tauri::State<'_, AppState>,
) -> AppResult<SpendingBreakdown> {
    finsight_api::commands::transactions::get_spending_breakdown(&state.api).await
}

#[tauri::command]
#[specta::specta]
pub async fn list_categories_with_spending(
    state: tauri::State<'_, AppState>,
) -> AppResult<Vec<CategoryWithSpending>> {
    finsight_api::commands::transactions::list_categories_with_spending(&state.api).await
}

#[tauri::command]
#[specta::specta]
pub async fn list_rules_with_categories(
    state: tauri::State<'_, AppState>,
) -> AppResult<Vec<RuleWithCategory>> {
    finsight_api::commands::transactions::list_rules_with_categories(&state.api).await
}

#[tauri::command]
#[specta::specta]
pub async fn toggle_rule(
    state: tauri::State<'_, AppState>,
    id: String,
    enabled: bool,
) -> AppResult<()> {
    finsight_api::commands::transactions::toggle_rule(&state.api, id, enabled).await
}

#[tauri::command]
#[specta::specta]
pub async fn get_transaction_count(state: tauri::State<'_, AppState>) -> AppResult<i64> {
    finsight_api::commands::transactions::get_transaction_count(&state.api).await
}

#[tauri::command]
#[specta::specta]
pub async fn set_transaction_flags(
    state: tauri::State<'_, AppState>,
    id: String,
    is_reimbursable: bool,
    is_split: bool,
) -> AppResult<Transaction> {
    finsight_api::commands::transactions::set_transaction_flags(
        &state.api,
        id,
        is_reimbursable,
        is_split,
    )
    .await
}

/// Record the user's verdict on whether a transaction is a transfer between
/// their own accounts. Sticky: survives re-imports and categorizer re-runs.
#[tauri::command]
#[specta::specta]
pub async fn set_transaction_transfer(
    state: tauri::State<'_, AppState>,
    id: String,
    is_transfer: bool,
) -> AppResult<TransferVerdictResult> {
    finsight_api::commands::transactions::set_transaction_transfer(&state.api, id, is_transfer)
        .await
}

/// Apply a transfer verdict to every undecided transaction matching the
/// counterparty pattern returned by `set_transaction_transfer`. One decision
/// clears a whole person's e-transfer history from the review list.
#[tauri::command]
#[specta::specta]
pub async fn apply_transfer_verdict_to_similar(
    state: tauri::State<'_, AppState>,
    pattern: String,
    is_transfer: bool,
) -> AppResult<u32> {
    finsight_api::commands::transactions::apply_transfer_verdict_to_similar(
        &state.api,
        pattern,
        is_transfer,
    )
    .await
}

/// Record the user's 3-way verdict (transfer / settle-up / real spending) on
/// a transfer-review counterparty transaction. Sticky: survives re-imports
/// and categorizer re-runs.
#[tauri::command]
#[specta::specta]
pub async fn set_counterparty_verdict(
    state: tauri::State<'_, AppState>,
    id: String,
    verdict: CounterpartyVerdict,
) -> AppResult<Transaction> {
    finsight_api::commands::transactions::set_counterparty_verdict(&state.api, id, verdict).await
}

/// Apply one counterparty verdict to every undecided transaction matching a
/// counterparty pattern (from [`UnresolvedCounterpartyDto::pattern`] or
/// `TransferVerdictResult::similar_pattern`). One decision clears a whole
/// person's e-transfer history from the review list.
#[tauri::command]
#[specta::specta]
pub async fn apply_counterparty_verdict_to_similar(
    state: tauri::State<'_, AppState>,
    pattern: String,
    verdict: CounterpartyVerdict,
) -> AppResult<u32> {
    finsight_api::commands::transactions::apply_counterparty_verdict_to_similar(
        &state.api, pattern, verdict,
    )
    .await
}

/// The undecided transfer-review queue, grouped by counterparty for a
/// bulk-decision surface.
#[tauri::command]
#[specta::specta]
pub async fn list_unresolved_counterparties(
    state: tauri::State<'_, AppState>,
) -> AppResult<Vec<UnresolvedCounterpartyDto>> {
    finsight_api::commands::transactions::list_unresolved_counterparties(&state.api).await
}

#[tauri::command]
#[specta::specta]
pub async fn get_transaction_splits(
    state: tauri::State<'_, AppState>,
    transaction_id: String,
) -> AppResult<Vec<TransactionSplitDto>> {
    finsight_api::commands::transactions::get_transaction_splits(&state.api, transaction_id).await
}

#[tauri::command]
#[specta::specta]
pub async fn set_transaction_splits(
    state: tauri::State<'_, AppState>,
    transaction_id: String,
    splits: Vec<SplitInputDto>,
) -> AppResult<()> {
    finsight_api::commands::transactions::set_transaction_splits(
        &state.api,
        transaction_id,
        splits,
    )
    .await
}

/// Returns CSV content for transactions matching a filter; the caller
/// downloads it client-side (Blob + `<a download>`). No native file dialog
/// since Phase 4 — the desktop shell has no local command surface to host one.
#[tauri::command]
#[specta::specta]
pub async fn export_transactions_csv(
    state: tauri::State<'_, AppState>,
    filter: TxnFilterInput,
) -> AppResult<String> {
    finsight_api::commands::transactions::export_transactions_csv(&state.api, filter).await
}

/// Re-run the Copilot `search_transactions` query and return the matching
/// rows as CSV content; the caller downloads it client-side. Shares
/// `transactions::search` with the Copilot tool so the exported rows match
/// exactly what the card displayed. No native file dialog since Phase 4.
#[tauri::command]
#[specta::specta]
pub async fn export_search_transactions_csv(
    state: tauri::State<'_, AppState>,
    query: SearchTxnQueryInput,
) -> AppResult<String> {
    finsight_api::commands::transactions::export_search_transactions_csv(&state.api, query).await
}
